modules/tests/src/main/scala/Person.scala

```diff
@@ -4,10 +4,12 @@ package tests
 import scalaz.{ -\/, \/, \/- }
 import monocle.Iso
 
+import shapeless.syntax.singleton._
+
 final case class Person(name: String, role: Option[Role])
 sealed trait Role
-final case class User(active: Boolean, boss: Person) extends Role
-final case class Admin(rights: List[String])         extends Role
+final case class User(active: Boolean)       extends Role
```

Comment:
Is this forever out of reach, or just for now?

Reply:
This is for sure the next wall I will bang my head against :)

I don't think it is impossible, but it would imply coming up with some sort of y-combinator for functions returning our "self-reflexive fix-point over a higher-order functor", which doesn't sound easy as cake (> <)'.

In the event that we fail to find a way through that, I'm personally fine with dropping support for recursive data-types.